### PR TITLE
feat(payment): PAYPAL-3113 Create a class to work with cookies

### PR DESCRIPTION
**Pull request by bc-nick**
## What?

Create a class to work with cookies

## Why?

To work with cookies

## Testing / Proof

Manual testing
Jest tests

Screenshot 2024-01-09 at 16 29 10

@bigcommerce/team-checkout @bigcommerce/team-payments

### CODE REVIEW

File: packages/storage/src/cookie-storage.spec.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+        expect(CookieStorage.get('agent')).toBe('Smith');
+    });
+
+    it('remove value by name', () => {
+        jest.spyOn(CookieStorage, 'set').mockImplementation();
+
+        CookieStorage.remove('last_user');
+
+        expect(CookieStorage.set).toHaveBeenCalledWith('last_user', '', { expires: new Date(0) });
```
Comment (serhii-tkachenko) [inline packages/storage/src/cookie-storage.spec.ts:32]: It might be great to check if the cookie with name `last_user` does not exist

**Review by serhii-tkachenko — APPROVED** [review]
Nice work!

**Review by bc-dronov — APPROVED** [review]

**Review by andriiVitvitskyi1990 — APPROVED** [review]

**Comment by animesh1987** [comment]
cc @bc-donfran can you please take a look at this as well.

File: packages/storage/src/cookie-storage.spec.ts (hunk 1 above)
Comment (animesh1987) [reply packages/storage/src/cookie-storage.spec.ts:32]: Are we testing all the conditions in `set` method?

**Comment by animesh1987** [comment]
Can you please attach a screenshot for the test coverage of this file?

**Comment by bc-nick** [comment]
> Can you please attach a screenshot for the test coverage of this file?

@animesh1987 thanks! updated PR description with screenshot

File: packages/storage/src/cookie-storage.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
+interface CookieOptions {
+    expires?: Date;
+    path?: string;
+    domain?: string;
+    secure?: boolean;
+}
+
+export default class CookieStorage {
+    static get(name: string): string | null {
+        const cookieValue = `${encodeURIComponent(`${name}`)}=`;
```
[the remark this answers is not in the document]
Comment (bc-nick) [reply packages/storage/src/cookie-storage.ts]: @animesh1987 done

**Comment by bc-donfran** [comment]
Hey @bc-nick! I've added two comments above.

File: packages/storage/src/cookie-storage.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        if (path) {
+            cookieText += `; path=${path}`;
+        }
+
+        if (domain) {
+            cookieText += `; domain=${domain}`;
+        }
+
+        if (secure) {
+            cookieText += '; secure';
```
[the remark this answers is not in the document]
Comment (bc-nick) [reply packages/storage/src/cookie-storage.ts:48]: Secure option means that browsers which support the secure attribute will only send cookies with the secure attribute when the request is going to an HTTPS page

File: packages/storage/src/cookie-storage.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        return value;
+    }
+
+    static set(name: string, value: string, options: CookieOptions = { secure: true }) {
+        let cookieText = `${encodeURIComponent(name)}=${encodeURIComponent(value)}`;
+
+        const { expires, path, domain, secure } = options;
+
+        if (expires) {
+            cookieText += `; expires=${expires.toUTCString()}`;
```
[the remark this answers is not in the document]
Comment (bc-nick) [reply packages/storage/src/cookie-storage.ts:36]: I think it's limited to 400 days

File: packages/storage/src/cookie-storage.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bc-donfran) [reply packages/storage/src/cookie-storage.ts:48]: Is there a reason for not using httponly as well?

File: packages/storage/src/cookie-storage.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (bc-donfran) [reply packages/storage/src/cookie-storage.ts:36]: Is there a reason for this cookie to be up to 400 days?

File: packages/storage/src/cookie-storage.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bc-nick) [reply packages/storage/src/cookie-storage.ts:48]: By default `secure` always true. I add possibility to change this value

File: packages/storage/src/cookie-storage.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (bc-nick) [reply packages/storage/src/cookie-storage.ts:36]: I guess not. But this option must be dynamic

[the remark this answers is not in the document]
Comment (bc-donfran) [reply packages/storage/src/cookie-storage.ts:36]: Yup that is fine with me! I think the expiration will be set elsewhere right?

File: packages/storage/src/cookie-storage.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bc-donfran) [reply packages/storage/src/cookie-storage.ts:48]: Yes please! Unless we need Javascript to read the cookies I would prefer both the "Secure" and "HTTPOnly" flag as this cookie does have a bit of sensitive data and we don't want XSS stealing it if not necessary!

File: packages/storage/src/cookie-storage.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (bc-nick) [reply packages/storage/src/cookie-storage.ts:36]: Maybe i did not understand you correct. What did you mean when you asking about maximum time?
Did you mean that we should have a maximum range for setting cookies?

400 days is the maximum default period from the time the cookie was set

File: packages/storage/src/cookie-storage.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bc-nick) [reply packages/storage/src/cookie-storage.ts:48]: I will add HTTPOnly option. Thanks for your advice!

File: packages/storage/src/cookie-storage.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (bc-donfran) [reply packages/storage/src/cookie-storage.ts:36]: Hey Nick! I think from a security perspective 400 days is a bit excessive. Are we able to lower it down?

[the remark this answers is not in the document]
Comment (bc-nick) [reply packages/storage/src/cookie-storage.ts:36]: This is dynamic option. We can set any expiration time depending on functional needs. By default, I can set any expiration time for the cookie. For example let it be 1 hour, do you agree?

[the remark this answers is not in the document]
Comment (bc-donfran) [reply packages/storage/src/cookie-storage.ts:36]: Yup that is perfect! Just wanted to make sure that we had an expiration time in mind but 1 hour is a good time. Thank you @bc-nick

File: packages/storage/src/cookie-storage.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bc-nick) [reply packages/storage/src/cookie-storage.ts:48]: @bc-donfran as far as I understand, we can't set this attribute (HTTPOnly) using JS. This attribute can be set on the backend side, and we cannot retrieve cookies using JS if they have this attribute.

File: packages/storage/src/cookie-storage.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (bc-nick) [reply packages/storage/src/cookie-storage.ts:36]: I have updated the PR with the changes for the default expiration time

File: packages/storage/src/cookie-storage.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bc-donfran) [reply packages/storage/src/cookie-storage.ts:56]: That makes sense but with the default expiration it should be fine. Thank you Nick!

File: packages/storage/src/cookie-storage.spec.ts (hunk 1 above)
Comment (bc-nick) [reply packages/storage/src/cookie-storage.spec.ts:55]: I am testing all conditions in set method. You can see this in the new commits

**Review by animesh1987 — APPROVED** [review]

**Pull request merged**